app/src/org/commcare/android/database/user/models/SessionStateDescriptor.java

```diff
@@ -82,49 +83,44 @@ public String getSessionDescriptor() {
-        //TODO: Serialize into something more useful. I dunno. JSON/XML/Something
+    private static String createSessionDescriptor(CommCareSession session) {
         StringBuilder descriptor = new StringBuilder();
         for (StackFrameStep step : session.getFrame().getSteps()) {
-            descriptor.append(step.getType());
-            if (SessionFrame.STATE_COMMAND_ID.equals(step.getType())) {
-                descriptor.append(step.getId());
-            } else if (SessionFrame.STATE_DATUM_VAL.equals(step.getType()) || SessionFrame.STATE_DATUM_COMPUTED.equals(step.getType())) {
-                descriptor.append(step.getId()).append(" ").append(step.getValue());
-            } else if (SessionFrame.STATE_QUERY_REQUEST.equals(step.getType())) {
-                // for now, don't support restoring sessions that make remote server requests
-                descriptor.append( SessionFrame.STATE_QUERY_REQUEST);
-            } else if (SessionFrame.STATE_SYNC_REQUEST.equals(step.getType())) {
-                // for now, don't support restoring sessions that make remote server requests
-                descriptor.append(SessionFrame.STATE_SYNC_REQUEST);
+            String type = step.getType();
+            if (!(SessionFrame.STATE_QUERY_REQUEST.equals(type) ||
+                    SessionFrame.STATE_SYNC_REQUEST.equals(type))) {
+                // Skip adding remote server query/sync steps to the descriptor.
+                // They are hard to replay (requires serializing query results)
+                // and shouldn't be needed for incomplete forms
+                descriptor.append(type).append(" ");
+            }
+
+            if (SessionFrame.STATE_COMMAND_ID.equals(type)) {
+                descriptor.append(step.getId()).append(" ");
+            } else if (SessionFrame.STATE_DATUM_VAL.equals(type)
+                    || SessionFrame.STATE_DATUM_COMPUTED.equals(type)) {
+                descriptor.append(step.getId()).append(" ").append(step.getValue()).append(" ");
             }
-            descriptor.append(" ");
         }
         return descriptor.toString().trim();
     }
 
-    public void loadSession(CommCareSession session) {
+    public void loadSessionFromDescriptor(CommCareSession session) {
         String[] tokenStream = sessionDescriptor.split(" ");
 
         int current = 0;
         while (current < tokenStream.length) {
             String action = tokenStream[current];
             if (action.equals(SessionFrame.STATE_COMMAND_ID)) {
                 session.setCommand(tokenStream[++current]);
-            } else if (action.equals(SessionFrame.STATE_DATUM_VAL) || action.equals(SessionFrame.STATE_DATUM_COMPUTED)) {
+            } else if (action.equals(SessionFrame.STATE_DATUM_VAL) ||
+                    action.equals(SessionFrame.STATE_DATUM_COMPUTED)) {
                 session.setDatum(tokenStream[++current], tokenStream[++current]);
-            } else if (action.equals(SessionFrame.STATE_SYNC_REQUEST)
```

Comment:
Yes. This code was introduced by https://github.com/dimagi/commcare-android/pull/1059/files#diff-cd3d036d05167f3e910a1381ea329b91R121 (in 2.28) and wasn't well thought out on my part, so I think these new changes are the right thing now.